File: ios/Betty/Features/GroupDetail/UserHistorySheet.swift
import SwiftUI

/// Web `UserHistory`: one row per game — the member's bet if placed, else a "NO BET"
/// skipped row for games that have already started. Sorted by kickoff, scores hidden
/// pre-kickoff unless sneak peek, header counts only actual bets ("<N> BETS · <Σ> PTS").
struct UserHistorySheet: View {
    let groupID: Int
    let userID: String

    @Environment(AppEnvironment.self) private var env
    @Environment(ThemeStore.self) private var theme
    @Environment(\.dismiss) private var dismiss

    private var group: Group? { env.groupStore.byID(groupID) }
    private var member: Member? { group?.member(withUserID: userID) }
    private var peek: Bool { group?.allowSneakPeek ?? false }

    private var games: [Game] {
        guard let group else { return [] }
        return env.tournamentStore.detailsByID(group.tournamentID)?.games ?? []
    }

    private var entries: [GroupUserHistoryEntry] {
        GroupUserHistoryLogic.entries(bets: env.betStore.bets, userID: userID, games: games)
    }

    var body: some View {
        // Web `.modal__inner` sits on the surface token, not the page background.
        ZStack(alignment: .topTrailing) {
            theme.colors.surface.ignoresSafeArea()
            VStack(alignment: .leading, spacing: 0) {
                header
                ScrollView {
                    VStack(spacing: 0) {
                        if entries.isEmpty {
                            Text("★ NO BETS YET")
                                .kicker(theme.colors.textMuted)
                                .padding(.vertical, Space.xl)
                                .frame(maxWidth: .infinity)
                        }
                        ForEach(Array(entries.enumerated()), id: \.element.id) { index, entry in
                            UserHistoryBetRow(
                                entry: entry,
                                peek: peek,
                                isMine: entry.bet.map { env.userStore.id == $0.userID } ?? false,
                                exactResultPoints: group?.exactResultPoints
                            )
                            .accessibilityElement(children: .combine)
                            .accessibilityIdentifier("groupDetail.userHistory.row.\(index)")
                            if entry.id != entries.last?.id {
                                Divider().overlay(theme.colors.overlay06)
                            }
                        }
                    }
                    .padding(.horizontal, 14)
                    .padding(.bottom, Space.l)
                }
            }
            closeButton
        }
        .accessibilityElement(children: .contain)
        .accessibilityIdentifier("groupDetail.userHistory.root")
        .presentationDetents([.medium, .large])
        .presentationDragIndicator(.visible)
        .task {
            guard let group else { return }
            if env.betStore.loadedGroupID != groupID {
                try? await env.betStore.load(groupID: groupID)
            }
            _ = try? await env.tournamentStore.loadDetails(id: group.tournamentID)
        }
    }

    private var closeButton: some View {
        Button {
            dismiss()
        } label: {
            Image(systemName: "xmark")
                .font(.system(size: 15, weight: .bold))
                .foregroundStyle(theme.colors.textSecondary)
                .padding(Space.s)
        }
        .padding(.top, Space.s)
        .padding(.trailing, Space.xs)
        .accessibilityLabel("Close")
    }

    private var header: some View {
        HStack(spacing: Space.m) {
            if let member {
                AvatarView(member: member, size: .medium)
            }
            VStack(alignment: .leading, spacing: 6) {
                Text("★ BET HISTORY").kicker(Palette.orange)
                Text((member?.displayName ?? "").uppercased())
                    .font(.bettyTitle2)
                    .foregroundStyle(theme.colors.textPrimary)
                    .lineLimit(1)
                    .truncationMode(.tail)
                HStack(spacing: Space.xs) {
                    Text("\(GroupUserHistoryLogic.betsCount(entries)) BETS")
                        .kicker(theme.colors.textSecondary)
                    Text("·").foregroundStyle(theme.colors.textMuted)
                    Text("\(GroupUserHistoryLogic.totalPoints(entries)) PTS")
                        .kicker(theme.colors.accentPositive)
                }
            }
            Spacer()
        }
        .padding([.top, .horizontal], Space.xl)
        .padding(.bottom, Space.l)
    }
}

/// Web `UserBetListItem`: team flags, the bet's score (hidden pre-kickoff unless peek —
/// but you always see your own), and a points chip once processed; result coloring uses
/// the group's exact points (legacy 3-or-4 heuristic when unknown). A `nil` entry.bet
/// renders a muted "NO BET" skipped row for games the member sat out.
struct UserHistoryBetRow: View {
    let entry: GroupUserHistoryEntry
    let peek: Bool
    let isMine: Bool
    let exactResultPoints: Int?

    @Environment(AppEnvironment.self) private var env
    @Environment(ThemeStore.self) private var theme

    private var showScore: Bool {
        guard let bet = entry.bet else { return false }
        return GroupBetRowLogic.showScore(bet: bet, gameStart: entry.game.startDate, peek: peek)
    }

    private var result: GroupBetRowLogic.Result {
        guard let bet = entry.bet else { return .pending }
        return GroupBetRowLogic.result(bet: bet, showScore: showScore, exactResultPoints: exactResultPoints)
    }

    var body: some View {
        HStack(spacing: 14) {
            HStack(spacing: 6) {
                TeamLogoView(team: env.teamStore.byID(entry.game.homeTeamID), size: 28)
                Text("–")
                    .font(.betty(14, .regular))
                    .foregroundStyle(theme.colors.textSecondary)
                TeamLogoView(team: env.teamStore.byID(entry.game.awayTeamID), size: 28)
            }

            HStack(alignment: .firstTextBaseline, spacing: 4) {
                if let bet = entry.bet {
                    if showScore || isMine {
                        Text("\(bet.homeTeamScore)")
                        Text("–")
                            .font(.betty(14, .regular))
                            .foregroundStyle(theme.colors.textSecondary)
                        Text("\(bet.awayTeamScore)")
                    } else {
                        HiddenScoreView()
                    }
                } else {
                    Text("NO BET").kicker(theme.colors.textSecondary)
                }
            }
            .font(.betty(18, .black))
            .monospacedDigit()
            .foregroundStyle(theme.colors.textPrimary)
            .frame(maxWidth: .infinity)

            pointsColumn
                .frame(width: 64, alignment: .trailing)
        }
        .padding(.vertical, 14)
        .padding(.horizontal, 12)
        .opacity(entry.bet == nil ? 0.55 : 1)
    }

    @ViewBuilder
    private var pointsColumn: some View {
        // Points stay pending until visible AND processed (your own pre-kickoff bet shows
        // its score but no points — isMine does not unlock points; web pin).
        if entry.bet == nil {
            Text("—")
                .font(.betty(18, .heavy))
                .foregroundStyle(theme.colors.textSecondary)
        } else if showScore, let bet = entry.bet, bet.isProcessed {
            let points = bet.userPoints ?? 0
            Text(points > 0 ? "+\(points)P" : "0P")
                .font(.bettyKicker)
                .kerning(0.8)
                .monospacedDigit()
                .foregroundStyle(chipColor)
                .padding(.vertical, 4)
                .padding(.horizontal, 8)
                .background(chipColor.opacity(0.15), in: RoundedRectangle(cornerRadius: Radius.sharp))
        } else {
            Text("·")
                .font(.betty(18, .heavy))
                .foregroundStyle(theme.colors.textSecondary)
        }
    }

    private var chipColor: Color {
        switch result {
        case .exact: theme.colors.accentPositive
        case .win: Palette.yellow
        case .miss: Palette.orange
        case .pending: theme.colors.textSecondary
        }
    }
}

File: android/app/src/main/java/social/betty/features/groupdetail/UserHistorySheet.kt
package social.betty.features.groupdetail

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import social.betty.core.model.Bet
import social.betty.core.model.Team
import social.betty.designsystem.BettyTheme
import social.betty.designsystem.Palette
import social.betty.designsystem.Radius
import social.betty.designsystem.Space
import social.betty.designsystem.components.Avatar
import social.betty.designsystem.components.AvatarSize
import social.betty.designsystem.components.HiddenScore
import social.betty.designsystem.components.TeamLogo
import social.betty.navigation.LocalAppContainer
import java.time.Instant

/**
 * Web `UserHistory`: one row per game — the member's bet if placed, else a "NO BET"
 * skipped row for games that have already started. Sorted by kickoff, scores hidden
 * pre-kickoff unless sneak peek, header counts only actual bets ("<N> BETS · <Σ> PTS").
 */
@Composable
fun UserHistorySheet(groupId: Int, userId: String, onDismiss: () -> Unit) {
    val container = LocalAppContainer.current
    val colors = BettyTheme.colors
    val type = BettyTheme.type
    val now = remember { Instant.now() }

    val groups by container.groupStore.groups.collectAsStateWithLifecycle()
    val group = groups.firstOrNull { it.id == groupId }
    val member = group?.members?.firstOrNull { it.userId == userId }
    val peek = group?.allowSneakPeek ?: false
    val myId = container.userStore.id

    val tournamentDetails by container.tournamentStore.details.collectAsStateWithLifecycle()
    val games = group?.let { tournamentDetails[it.tournamentId]?.games } ?: emptyList()

    val teams by container.teamStore.teams.collectAsStateWithLifecycle()
    fun teamBy(id: Int): Team? = teams.firstOrNull { it.id == id }

    var bets by remember { mutableStateOf<List<Bet>>(emptyList()) }
    val entries = remember(bets, userId, games, now) {
        GroupUserHistoryLogic.entries(bets, userId, games, now)
    }

    LaunchedEffect(groupId) {
        group?.let { runCatching { container.tournamentStore.loadDetails(it.tournamentId) } }
        if (teams.isEmpty()) runCatching { container.teamStore.load() }
        runCatching { container.api.getBetsByGroup(groupId) }.onSuccess { bets = it }
    }

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .background(colors.surface)
            .testTag("user-history-sheet"),
    ) {
        // Header.
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(top = Space.l, start = Space.xl, end = Space.xl, bottom = Space.l),
            horizontalArrangement = Arrangement.spacedBy(Space.m),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            if (member != null) {
                Avatar(url = member.imageUrl, name = member.displayName(), size = AvatarSize.medium)
            }
            Column(verticalArrangement = Arrangement.spacedBy(6.dp), modifier = Modifier.weight(1f)) {
                Text("★ BET HISTORY", style = type.kicker, color = Palette.orange)
                Text(
                    text = (member?.displayName() ?: "").uppercase(),
                    style = type.title2,
                    color = colors.textPrimary,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
                Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(Space.xs)) {
                    Text(
                        text = "${GroupUserHistoryLogic.betsCount(entries)} BETS",
                        style = type.kicker,
                        color = colors.textSecondary,
                    )
                    Text("·", color = colors.textMuted)
                    Text(
                        text = "${GroupUserHistoryLogic.totalPoints(entries)} PTS",
                        style = type.kicker,
                        color = colors.accentPositive,
                    )
                }
            }
        }

        Column(
            modifier = Modifier
                .fillMaxWidth()
                .heightIn(max = 480.dp)
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 14.dp)
                .padding(bottom = Space.l),
        ) {
            if (entries.isEmpty()) {
                Text(
                    text = "★ NO BETS YET",
                    style = type.kicker,
                    color = colors.textMuted,
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(vertical = Space.xl),
                    textAlign = TextAlign.Center,
                )
            }
            entries.forEachIndexed { index, entry ->
                UserHistoryBetRow(
                    entry = entry,
                    peek = peek,
                    isMine = entry.bet?.let { myId == it.userId } ?: false,
                    exactResultPoints = group?.exactResultPoints,
                    homeTeam = teamBy(entry.game.homeTeamId),
                    awayTeam = teamBy(entry.game.awayTeamId),
                    now = now,
                )
                if (index != entries.lastIndex) {
                    Box(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(horizontal = 4.dp)
                            .height(1.dp)
                            .background(colors.overlay06),
                    )
                }
            }
        }
    }
}

@Composable
private fun UserHistoryBetRow(
    entry: UserHistoryEntry,
    peek: Boolean,
    isMine: Boolean,
    exactResultPoints: Int?,
    homeTeam: Team?,
    awayTeam: Team?,
    now: Instant,
) {
    val colors = BettyTheme.colors
    val type = BettyTheme.type
    val bet = entry.bet
    val showScore = bet != null && GroupBetRowLogic.showScore(bet, entry.game.startDate, peek, now)
    val result = bet?.let { GroupBetRowLogic.result(it, showScore, exactResultPoints) }
        ?: GroupBetRowLogic.Result.PENDING

    Row(
        modifier = Modifier
            .fillMaxWidth()
            .alpha(if (bet == null) 0.55f else 1f)
            .padding(vertical = 14.dp, horizontal = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(14.dp),
    ) {
        Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(6.dp)) {
            TeamLogo(url = homeTeam?.imageUrl, name = homeTeam?.name, size = 28.dp)
            Text("–", style = type.bodyRegular, color = colors.textSecondary)
            TeamLogo(url = awayTeam?.imageUrl, name = awayTeam?.name, size = 28.dp)
        }

        Box(modifier = Modifier.weight(1f), contentAlignment = Alignment.Center) {
            if (bet == null) {
                Text("NO BET", style = type.kicker, color = colors.textSecondary)
            } else if (showScore || isMine) {
                Text(
                    text = "${bet.homeTeamScore} – ${bet.awayTeamScore}",
                    style = type.score.copy(fontSize = 18.sp),
                    color = colors.textPrimary,
                )
            } else {
                HiddenScore()
            }
        }

        Box(modifier = Modifier.width(64.dp), contentAlignment = Alignment.CenterEnd) {
            // Points stay pending until visible AND processed (isMine does not unlock points).
            if (bet == null) {
                Text("—", style = type.subhead, color = colors.textSecondary)
            } else if (showScore && bet.isProcessed) {
                val points = bet.userPoints ?: 0
                val chipColor = when (result) {
                    GroupBetRowLogic.Result.EXACT -> colors.accentPositive
                    GroupBetRowLogic.Result.WIN -> Palette.yellow
                    GroupBetRowLogic.Result.MISS -> Palette.orange
                    GroupBetRowLogic.Result.PENDING -> colors.textSecondary
                }
                Text(
                    text = if (points > 0) "+${points}P" else "0P",
                    style = type.kicker,
                    color = chipColor,
                    modifier = Modifier
                        .clip(Radius.sharp)
                        .background(chipColor.copy(alpha = 0.15f))
                        .padding(vertical = 4.dp, horizontal = 8.dp),
                )
            } else {
                Text("·", style = type.subhead, color = colors.textSecondary)
            }
        }
    }
}
